akka-actor-typed/src/main/scala/akka/actor/typed/eventstream/EventStream.scala

```diff
@@ -0,0 +1,70 @@
+/*
+ * Copyright (C) 2019 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package akka.actor.typed.eventstream
+
+import akka.actor.typed._
+import akka.annotation.{ DoNotInherit, InternalApi }
+
+import scala.reflect.ClassTag
+
+object EventStream {
+
+  /**
+   * Not for user Extension
+   */
+  @DoNotInherit sealed trait Command
+
+  /**
+   * Publish an event of type E
+   * @param event
+   * @tparam E
+   */
+  final case class Publish[E](event: E) extends Command
+
+  /**
+   * Subscribe a typed actor to listen for types or subtypes of E.
+   * ==Simple example==
+   * {{{
+   *   sealed trait A
+   *   case object A1 extends A
+   *   //listen for all As
+   *   def subscribe(actorSystem: ActorSystem[_], actorRef: ActorRef[A]) =
+   *     actorSystem.eventStream ! Subscribe(actorRef)
+   *   //listen for A1s only
+   *   def subscribe(actorSystem: ActorSystem[_], actorRef: ActorRef[A]) =
+   *     actorSystem.eventStream ! Subscribe[A1](actorRef)
+   * }}}
+   *
+   * @param subscriber
+   * @param classTag
+   * @tparam E
+   */
+  final case class Subscribe[E](subscriber: ActorRef[E])(implicit classTag: ClassTag[E]) extends Command {
+
+    /**
+     * Java API.
+     */
+    def this(subscriber: ActorRef[E], clazz: Class[E]) = this(subscriber)(ClassTag(clazz))
+
+    /**
+     * INTERNAL API
+     */
+    @InternalApi private[akka] def topic: Class[_] = classTag.runtimeClass
+  }
+
+  /**
+   * Unsubscribe an actor ref from the event stream
+   * @param subscriber
+   * @tparam E
+   */
+  final case class Unsubscribe[E](subscriber: ActorRef[E]) extends Command
+
+  /**
+   * Java API.
+   */
+  def subscribeToType[E](subscriber: ActorRef[E], clazz: java.lang.Class[E]): Subscribe[E] =
```

Comment:
I don't see that this adds much more value over `new EventStream.Subscribe` and can just be confusing to have several ways. As mentioned in my original comment this also suffers from that it's not clear that this is just a factory method (without side effects).

Best would be to have such factory method as a static method in the `Subscribe` class.

`Subscribe.create` or `Subscribe.of`

However, then we can't use nested Scala objects `EventStream.Subscribe`. Here we have a separate package for `eventstream` so maybe it's better to drop the `EventStream` object and place the messages at the top. Then we could also have the factory methods at the "right" place.

Wdyt?

Reply:
> However, then we can't use nested Scala objects EventStream.Subscribe. Here we have a separate package for eventstream so maybe it's better to drop the EventStream object and place the messages at the top. Then we could also have the factory methods at the "right" place.

Yes, actually this makes more sense, will do